cmd/otel-allocator/server/server.go

```diff
@@ -0,0 +1,189 @@
+	router.HandleFunc("/jobs/{job_id}/targets", s.TargetsHandler).Methods("GET")
+	router.Path("/metrics").Handler(promhttp.Handler())
+	s.server = &http.Server{Addr: *listenAddr, Handler: router, ReadHeaderTimeout: 90 * time.Second}
+	return s
+}
+
+func (s *Server) Start() error {
+	s.logger.Info("Starting server...")
+	return s.server.ListenAndServe()
+}
+
+func (s *Server) Shutdown(ctx context.Context) error {
+	s.logger.Info("Shutting down server...")
+	return s.server.Shutdown(ctx)
+}
+
+// ScrapeConfigsHandler returns the available scrape configuration discovered by the target allocator.
+// The target allocator first marshals these configurations such that the underlying prometheus marshaling is used.
+// After that, the YAML is converted in to a JSON format for consumers to use.
+func (s *Server) ScrapeConfigsHandler(w http.ResponseWriter, r *http.Request) {
+	configs := s.discoveryManager.GetScrapeConfigs()
+
+	hash, err := hashstructure.Hash(configs, nil)
+	if err != nil {
+		s.logger.Error(err, "failed to hash the config")
+		s.errorHandler(w, err)
+		return
+	}
+	// if the hashes are different, we need to recompute the scrape config
+	if hash != s.compareHash {
+		var configBytes []byte
+		configBytes, err = yaml.Marshal(configs)
+		if err != nil {
+			s.errorHandler(w, err)
+			return
+		}
+		var jsonConfig []byte
+		jsonConfig, err = yaml2.YAMLToJSON(configBytes)
+		if err != nil {
+			s.errorHandler(w, err)
+			return
+		}
+		s.scrapeConfigResponse = jsonConfig
+		s.compareHash = hash
+	}
+	// We don't use the jsonHandler method because we don't want our bytes to be re-encoded
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(s.scrapeConfigResponse)
+	if err != nil {
+		s.errorHandler(w, err)
+	}
+}
+
+func (s *Server) JobHandler(w http.ResponseWriter, r *http.Request) {
+	displayData := make(map[string]target.LinkJSON)
+	for _, v := range s.allocator.TargetItems() {
+		displayData[v.JobName] = target.LinkJSON{Link: v.Link.Link}
+	}
+	s.jsonHandler(w, displayData)
+}
+
+// PrometheusMiddleware implements mux.MiddlewareFunc.
+func (s *Server) PrometheusMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		route := mux.CurrentRoute(r)
+		path, _ := route.GetPathTemplate()
+		timer := prometheus.NewTimer(httpDuration.WithLabelValues(path))
+		next.ServeHTTP(w, r)
+		timer.ObserveDuration()
+	})
+}
+
+func (s *Server) TargetsHandler(w http.ResponseWriter, r *http.Request) {
+	q := r.URL.Query()["collector_id"]
+
+	params := mux.Vars(r)
+	jobId, err := url.QueryUnescape(params["job_id"])
+	if err != nil {
+		s.errorHandler(w, err)
+		return
+	}
+
+	if len(q) == 0 {
+		displayData := GetAllTargetsByJob(s.allocator, jobId)
+		s.jsonHandler(w, displayData)
+
+	} else {
+		tgs := s.allocator.GetTargetsForCollectorAndJob(q[0], jobId)
+		// Displays empty list if nothing matches
+		if len(tgs) == 0 {
```

Comment:
Hmm wondering if L164 would also cover this case?  Isn't tgs already an empty list in this case, so could you display it directly?